File: snapshots/input/syntax/Main/Expressions.cs
using System.Diagnostics.CodeAnalysis;

namespace Main;

[SuppressMessage("ReSharper", "all")]
public class Expressions
{
    void AssignmentToPrefixUnaryExpressions()
    {
        var a = 42;
        var b = 42;
        a = +a;
        a = -a;
        a = ~a;
        a = ++a;
        a = --a;
        a = a++;
        a = a--;
        b = a!;

        var c = true;
        c = !c;
    }

    void AssignmentToPrefixBinaryExpressions()
    {
        var a = 42;
        a = a + a;
        a = a - a;
        a = a * a;
        a = a / a;
        a = a % a;
        a = a & a;
        a = a | a;
        a = a ^ a;
        a = a >> a;
        a = a << a;
        a = a >>> a;
    }

    void AssignmentToBinaryEqualityExpression()
    {
        var a = true;
        var b = true;
        var c = 42;
        var d = 42;
        a = a == b;
        a = a != b;
        a = c < d;
        a = c <= d;
        a = c > d;
        a = c >= d;
    }

    void AssignmentToBinaryExpression()
    {
        var a = 42;
        a += a;
        a -= a;
        a *= a;
        a /= a;
        a %= a;
        a++;
        a--;
        a <<= a;
        a >>= a;
        a >>>= a;
    }

    struct Struct
    {
        public int Property;
    }

    struct IndexedClass
    {
        public int Property;

        public int this[int index]
        {
            get { return Property; }
            set { Property = value; }
        }
    }

    void AssignmentToLeftValueTypes()
    {
        (var a, var b) = (1, 2);
        a = 1;
        var c = new Struct { Property = 42 };
        c.Property = 1;
        var d = new IndexedClass();
        d[b] = 1;
        (a, b) = (1, 2);
        var x = new IndexedClass
        {
            Property = 1,
            [b] = 1
        };
        (a) = 1;
        unsafe
        {
            int myInt = 5;
            int* p = &myInt;
            Console.WriteLine("myInt = {0}, *p = {1}", myInt, *p);
        }
    }

    void TernaryExpression()
    {
        var x = true;
        var y = x ? "foo" : "bar";
        object z = true;
        var t = z is bool ? 42 : 41;
    }

    class Cast
    {
        public Cast nested;
        public Cast2 nested2;

        public Cast plus(Cast other)
        {
            nested = other;
            return this;
        }

        public class Cast2
        {
        }
    }

    int CastExpressions()
    {
        object a = new Cast();
        object b = new Cast();
        Cast c = ((Cast)a).plus((Cast)b);
        Cast d = (Cast)new object[] { a, b }[0];
        var e = (Cast.Cast2)(c.nested.nested2);
        var f = (Int32)(1);
        var g = (Int32)(1);
        var h = (Int32)((1));
        return f + g + h;
    }

    object AnonymousObject()
    {
        var x = new { Helper = "" };
        var y = new
        {
            x
        };
        return y.x.Helper;
    }

    class TargetType
    {
        public TargetType(string name)
        {
        }
    }

    TargetType TargetTypeNew()
    {
        TargetType x = new("x");
        return x;
    }

    int Checked()
    {
        var three = checked(1 + 2);
        return three;
    }

    class ObjectCreationClass
    {
        public D field;

        public ObjectCreationClass(D field)
        {
            this.field = field;
        }

        public class D
        {
            public D(int a, string b)
            {
            }
        }
    }

    void ObjectCreation()
    {
        var a = new ObjectCreationClass.D(1, "hi");
        var b = new ObjectCreationClass(a)
        {
            field = a,
        };
        b = new ObjectCreationClass(a);
        b = new ObjectCreationClass(a) { };
    }

    class NamedParametersClass
    {
        public int A;
        public string B;

        public NamedParametersClass(int a, string b)
        {
            A = a;
            B = b;
        }

        public void Update(int a, string b)
        {
            A = a;
            B = b;
        }
    }

    NamedParametersClass NamedParameters()
    {
        var a = new NamedParametersClass(b: "hi", a: 1);
        a.Update(b: "foo", a: 42);
        return a;
    }

    Func<int, int> AnonymousFunction()
    {
        var d = delegate (int _, int _) { return 42; };
        return delegate (int a) { return a + d.Invoke(a, a); };
    }

    class Lambda
    {
        public string func(Lambda x)
        {
            return "";
        }
    }

    void LambdaExpressions()
    {
        var a = (string x) => x + 1;
        var b = (Lambda a, Lambda b) => { return a.func(b); };
        var c = string (Lambda a, Lambda _) => { return a; };
    }

    void TupleExpressions()
    {
        var a = (1, 2, "");
    }

    void ArrayCreation()
    {
        var a = new[,] { { 1, 1 }, { 2, 2 }, { 3, 3 } };
        Span<int> b = stackalloc[] { 1, 2, 3 };
        Span<int> c = stackalloc int[] { 1, 2, 3 };
        var d = new int[3] { 1, 2, 3 };
        var e = new byte[,] { { 1, 2 }, { 2, 3 } };
        var f = new int[3, 2] { { 1, 1 }, { 2, 2 }, { 3, 3 } };
        var g = new (string b, string c)[3];
    }

    void MakeRef()
    {
        var g = "";
        var a = __makeref(g);
    }

    void SizeOf()
    {
        var a = sizeof(int);
    }

    void TypeOf()
    {
        var a = typeof(int);
        var b = typeof(List<string>.Enumerator);
        var c = typeof(Dictionary<,>);
        var d = typeof(Tuple<,,,>);
    }

    interface IAnimal
    {
        string Sound();
    }

    public class Dog : IAnimal
    {
        public string Sound()
        {
            return "woof";
        }
    }

    public class Cat : IAnimal
    {
        public string Sound()
        {
            return "meow";
        }
    }

    void Switch()
    {
        int some = 42;
        var a = some switch
        {
            1 => "one",
            2 => "two",
            _ => "more"
        };
        IAnimal dog = new Dog();
        var b = dog switch
        {
            Cat c => c.Sound(),
            Dog c => c.Sound(),
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    void Dictionary()
    {
        var a = new Dictionary<string, int> { ["a"] = 65 };
    }

    void Is()
    {
        object s = "s";
        if (s is string s2)
        {
            Console.WriteLine(s2);
        }

        var c = s is "test";
        var a = s is int.MaxValue;
        var d = s is nameof(a);
    }
}

File: snapshots/input/syntax/VBMain/Expressions.vb
Imports System.Diagnostics.CodeAnalysis

Namespace VBMain
    <SuppressMessage("ReSharper", "all")>
    Public Class Expressions

        Private Sub AssignmentToPrefixUnaryExpressions()
            Dim A = 42
            Dim B = 42
            A = +A
            A = -A
            A = Not A
            B = A
            Dim C = True
            C = Not C
        End Sub

        Private Sub AssignmentToPrefixBinaryExpressions()
            Dim A = 42
            A = A + A
            A = A - A
            A = A * A
            A = A / A
            A = A \ A
            A = A ^ A
            A = A Mod A
            A = A And A
            A = A Or A
            A = A Xor A
            A = A << A
            A = A >> A
        End Sub

        Private Sub AssignmentToBinaryEqualityExpression()
            Dim A = True
            Dim B = True
            Dim C = 42
            Dim D = 42
            A = A = B
            A = A <> B
            A = C < D
            A = C <= D
            A = C > D
            A = C >= D
        End Sub

        Private Sub AssignmentToBinaryExpression()
            Dim A = 42
            A += A
            A -= A
            A *= A
            A /= A
            A \= A
            A &= A
            A <<= A
            A >>= A
            A ^= A
        End Sub

        Structure Struct
            Public [Property] As Integer
        End Structure

        Structure IndexedClass
            Public [Property] As Integer

            Default Public Property Item(ByVal index As Integer) As Integer
                Get
                    Return [Property]
                End Get
                Set(ByVal value As Integer)
                    [Property] = value
                End Set
            End Property
        End Structure

        Private Sub AssignmentToLeftValueTypes()
            Dim E As (A As Integer, B As Integer) = (1, 2)
            Dim A = 1
            Dim C = New Struct With {
                .[Property] = 42
            }
            C.[Property] = 1
            Dim D = New IndexedClass()
            D(E.B) = 1
            Dim X = New IndexedClass With {
                .[Property] = 1
            }
            E.A = 1
        End Sub

        Private Sub TernaryExpression()
            Dim X = True
            Dim Y = If(X, "foo", "bar")
            Dim Z As Object = True
            Dim T = If(TypeOf Z Is Boolean, 42, 41)
        End Sub

        Class Cast
            Public Nested As Cast
            Public Nested2 As Cast2

            Public Function Plus(ByVal other As Cast) As Cast
                Nested = other
                Return Me
            End Function

            Public Class Cast2
            End Class
        End Class

        Private Function CastExpressions() As Integer
            Dim A As Object = New Cast()
            Dim B As Object = New Cast()
            Dim C As Cast = (CType(A, Cast)).Plus(CType(B, Cast))
            Dim D As Cast = CType(New Object() {A, B}(0), Cast)
            Dim E = CType((C.Nested.Nested2), Cast.Cast2)
            Dim F = CType((1), Int32)
            Dim G = CType((1), Int32)
            Dim H = CType(((1)), Int32)
            Return F + G + H
        End Function

        Private Function AnonymousObject() As Object
            Dim X = New With {Key .Helper = ""}
            Dim Y = New With {X}
            Return Y.x.Helper
        End Function

        Class ObjectCreationClass
            Public Field As D

            Public Sub New(ByVal field As D)
                Me.Field = field
            End Sub

            Public Class D
                Public Sub New(ByVal a As Integer, ByVal b As String)
                End Sub
            End Class
        End Class

        Private Sub ObjectCreation()
            Dim A = New ObjectCreationClass.D(1, "hi")
            Dim B = New ObjectCreationClass(A) With {
                .Field = A
            }
            B = New ObjectCreationClass(A)
        End Sub

        Class NamedParametersClass
            Public A As Integer
            Public B As String

            Public Sub New(ByVal a As Integer, ByVal b As String)
                Me.A = a
                Me.B = b
            End Sub

            Public Sub Update(ByVal a As Integer, ByVal b As String)
                Me.A = a
                Me.B = b
            End Sub
        End Class

        Private Function NamedParameters() As NamedParametersClass
            Dim A = New NamedParametersClass(b:="hi", a:=1)
            A.Update(b:="foo", a:=42)
            Return A
        End Function

        Private Function AnonymousFunction() As Func(Of Integer, Integer)
            Dim d = Function(ByVal __ As Integer, ByVal ___ As Integer) 42
            Return Function(ByVal a As Integer) a + d.Invoke(a, a)
        End Function

        Class Lambda
            Public Function func(ByVal x As Lambda) As String
                Return ""
            End Function
        End Class

        Private Sub LambdaExpressions()
            Dim a = Function(ByVal x As String) x & 1
            Dim b = Function(ByVal aa As Lambda, ByVal bb As Lambda) aa.func(bb)
            Dim c = Function(aaa As Lambda, __ As Lambda)
                        If True Then
                            Return "hi"
                        End If
                    End Function
        End Sub

        Private Sub TupleExpression()
            Dim A = (1, 2, "")
        End Sub

        Private Sub ArrayCreation()
            Dim a = {
            {1, 1},
            {2, 2},
            {3, 3}}
            Dim d = New Integer(2) {1, 2, 3}
            Dim e = New Byte(,) {
            {1, 2},
            {2, 3}}
            Dim f = New Integer(2, 1) {
            {1, 1},
            {2, 2},
            {3, 3}}

            Dim numbers(4) As Integer
            Dim numbers2 = New Integer() {1, 2, 4, 8}
            ReDim Preserve numbers(15)
            ReDim numbers(15)
            Dim matrix(5, 5) As Double
            Dim matrix2 = New Integer(,) {{1, 2, 3}, {2, 3, 4}, {3, 4, 5}, {4, 5, 6}}
            Dim sales()() As Double = New Double(11)() {}
        End Sub

        Private Sub [TypeOf]()
            Dim a = GetType(Integer)
            Dim b = GetType(List(Of String).Enumerator)
            Dim c = GetType(Dictionary(Of,))
            Dim d = GetType(Tuple(Of,,,))
        End Sub

        Private Sub SelectCase()
            Dim Some = 42
            Select Case Some
                Case 1
                    Debug.WriteLine("One")
                Case 2
                    Debug.WriteLine("One")
                Case Else
                    Debug.WriteLine("More")
            End Select
        End Sub

        Private Sub Dictionary()
            Dim A = New Dictionary(Of String, Integer) From {{1, "Test1"}, {2, "Test1"}}
        End Sub

    End Class
End Namespace
